### PR TITLE
ci: progress on codesigning macOS builds

### PR DESCRIPTION
**Pull request by ErikBjare**
Progress on code signing, building on lessons learned in #344.

Co-authored-by: @xylix

## What works

 - pyinstaller produces a signed `.app`, which passes `xcnotary precheck` and doesn't return 'adhoc' when checked with `codesign -cv ActivityWatch.app`.

# Issues

 - No longer opens when double-clicking the `.app`, might be due to https://github.com/ActivityWatch/aw-qt/pull/76
   - We've had this before, can't remember the cause, but something about relative working dirs?
   - Might also be due to the app being signed but not notarized?
   - Might be due to https://github.com/pyinstaller/pyinstaller/issues/1892 (crash when writing default config)
     - Potential workaround in https://github.com/pyinstaller/pyinstaller/issues/5109#issuecomment-683681106
 - ~~Accessibility perms don't seem to work anymore? Always getting 'unknown' events.~~
   - Getting "Not authorized to send Apple events to System Events" errors when using the 'applescript' strategy.
     - Apparently you need to add an exception to the Info.plist...
     - The entitlements don't seem to be applied by PyInstaller, applying them manually with `codesign -f -s $SIGNER --entitlement entitlement.plist ActivityWatch.app/Contents/MacOS/aw-watcher-window` to aw-qt and aw-watcher-window after the build is done seems to work, at least for the 'applescript' strategy.
   - Might be due to hardened runtime?
 - [x] CI needs access to the signing key (see #344)
   - ~~Should use https://github.com/FiloSottile/age to securely deliver the key (instead of the complicated openssl command used in #344)~~
   - [ ] If no key present, should skip or use ad-hoc signing.

# Related PRs

 - https://github.com/ActivityWatch/activitywatch/pull/344
 - https://github.com/ActivityWatch/activitywatch/pull/699

# Bounty recipients

 - @kewde provided a lot of help in #699, and is eligible for a part of the bounty
 - @xylix provided help in the past in #344, and is eligible for a part of the bounty

# Resources

 - https://github.com/pyinstaller/pyinstaller/pull/5581
 - https://github.com/pyinstaller/pyinstaller/pull/3991
 - https://github.com/akeru-inc/xcnotary
 - https://stackoverflow.com/questions/64652704/how-to-notarize-an-macos-command-line-tool-created-outside-of-xcode

### CODE REVIEW

File: aw.spec [hunk 1]
```diff
@@ -17,6 +17,9 @@ current_release = subprocess.run(
 ).stdout.strip()
 print("bundling activitywatch version " + current_release)
 
+entitlements_file = Path(".") / "scripts" / "package" / "entitlements.plist"
```
Comment (ErikBjare) [inline aw.spec:20]: Using `.` here is probably bad.

**Comment by oreHGA** [comment]
Happy holidays! Was trying the latest version on the Mac OS Monterey 12.0.1 Screen Shot 2021-12-31 at 12 39 51 AM

Would this be something that this PR fixes? 
Screen Shot 2021-12-31 at 12 39 51 AM

**Comment by kewde** [comment]
I managed to get it to build but notarization will not succeed (neither on the app or the dmg).
The hardened runtime is not enabled (`flags=0x2(adhoc)` should be `flags=0x10000(runtime)`):
```
(venv) user@MacOSs-iMac dist % codesign -dvv ActivityWatch.app
Executable=/Users/user/Documents/projects/activitywatch/activitywatch/dist/ActivityWatch.app/Contents/MacOS/aw-qt
Identifier=net.activitywatch.ActivityWatch
Format=app bundle with Mach-O thin (x86_64)
CodeDirectory v=20400 size=20920 flags=0x2(adhoc) hashes=643+7 location=embedded
Signature=adhoc
Info.plist entries=12
TeamIdentifier=not set
Sealed Resources version=2 rules=13 files=992
Internal requirements count=0 size=12
```
https://wiki.freepascal.org/Hardened_runtime_for_macOS#Verifying_hardening

**Comment by ErikBjare** [comment]
@kewde Thanks for the reply, noting the same issue. In the CI logs I see a couple lines of `Code signing identity: None`, which suggests PyInstaller only does ad-hoc signing. 

I also found https://github.com/pyinstaller/pyinstaller/issues/4701 which suggests you can turn on hardened runtime at `codesign`-time with the `--options=runtime` param.

**Comment by kewde** [comment]
@ErikBjare Investigating this.

It shouldn't be adhoc signing, instead it should be signing using the specified identifier.

**Comment by ErikBjare** [comment]
Might be due to the certificate not being available when PyInstaller does its thing (if that is the case, would be nice if it threw an error instead).

So I guess we need to retrieve the cert somehow, I assume you can't do that with just `xcrun notarytool store-credentials ...`?

Also found this, which seems pretty comprehensive: https://stackoverflow.com/questions/64652704/how-to-notarize-an-macos-command-line-tool-created-outside-of-xcode

**Comment by kewde** [comment]
The certificate is available on my machine, so it should work. This is something else, will comment once I know.

**Comment by kewde** [comment]
@ErikBjare I used the original `codesign.sh` script which did do a proper job in signing, it enabled the hardened runtime and it set the team identifier to my personal team. So this is a PyInstaller issue - taking a deeper look!
```
user@MacOSs-iMac dist % codesign --verify --display --verbose ActivityWatch.app
Executable=/Users/user/Documents/projects/activitywatch/activitywatch/dist/ActivityWatch.app/Contents/MacOS/aw-qt
Identifier=net.activitywatch.ActivityWatch
Format=app bundle with Mach-O thin (x86_64)
CodeDirectory v=20500 size=20939 flags=0x10000(runtime) hashes=643+7 location=embedded
Signature size=8980
Timestamp=5 Jan 2022 at 12:26:22
Info.plist entries=12
TeamIdentifier=MYIDENTIFIER
Runtime Version=11.3.0
Sealed Resources version=2 rules=13 files=992
Internal requirements count=1 size=192
```

**Comment by ErikBjare** [comment]
Thanks for all the help on this @kewde. I don't have access to a macOS box today, but will investigate further tomorrow or friday.

**Comment by kewde** [comment]
@ErikBjare codesigning & notarization works, will PR :tada: 

PyInstaller codesigning _command line_ is rather buggy: the `--codesign-idenity` does not work.
The `codesign_identity` for the `BUNDLE(...)` is derived from the `EXE(...)` and `COLLECT(...)`. 

Notarization output
```
bash ./scripts/notarize.sh
Detecting availability of notarization tools
+ Found notarytool
Notarizing: dist/ActivityWatch.app

This process stores your credentials securely in the Keychain. You reference these credentials later using a profile name.

Validating your credentials...
Success. Credentials validated.
Credentials saved to Keychain.
To use them, specify `--keychain-profile "ActivityWatchSigner"`
Notarization: starting for dist/ActivityWatch.app.zip
Notarization: in progress for dist/ActivityWatch.app.zip
Conducting pre-submission checks for ActivityWatch.app.zip and initiating connection to the Apple notary service...
Submission ID received
  id: SNIPPED
Successfully uploaded file52,7 MB of 52,7 MB)   
  id: SNIPPED
  path: /Users/user/Documents/projects/activitywatch/activitywatch/dist/ActivityWatch.app.zip
Waiting for processing to complete.
Current status: Accepted...............
Processing complete
  id: SNIPPED
  status: Accepted
Processing: /Users/user/Documents/projects/activitywatch/activitywatch/dist/ActivityWatch.app
Processing: /Users/user/Documents/projects/activitywatch/activitywatch/dist/ActivityWatch.app
The staple and validate action worked!
```

**Comment by kewde** [comment]
@ErikBjare it seems like the CU is not importing the certificate into the keychain, fixed in #703.

Please note: for notarization you must use a developer id (used for distribution), not a development id.

**Comment by ErikBjare** [comment]
Seems like signing works now, nice!

Still can't get the app to open when I double-click in Finder, but works with `open ActivityWatch.app` from the terminal. That can wait for another PR though.

**Comment by ErikBjare** [comment]
Merging this, fixing the double-click in Finder will have to wait for another PR...

**Comment by ErikBjare** [comment]
Huge thanks @kewde for helping with getting this done!

**Comment by kewde** [comment]
@ErikBjare re: double clicking Finder
We should check out the `info.plist` file - it's autogenerated by PyInstaller but we can modify it a bit iirc.
https://github.com/ActivityWatch/activitywatch/issues/630#issuecomment-929400494

**Comment by ErikBjare** [comment]
@kewde Unfortunately it does not seem to be that simple. Pretty sure that the cause/fix is https://github.com/pyinstaller/pyinstaller/issues/5109#issuecomment-683313824. I was working briefly on a minimum viable reproducible example yesterday, and managed to use macOS `syslog` as a way to get some basic logging/output even when double-clicked from Finder.

**Pull request merged**